tests/test_shapely.py

```diff
@@ -0,0 +1,143 @@
+import pytest
+
+from psycopg.pq import Format
+from psycopg.types import TypeInfo
+
+from psycopg import ProgrammingError
+
+pytestmark = [pytest.mark.postgis]
+pytest.importorskip("shapely")
+
+# real example, with CRS and "holes"
+MULTIPOLYGON_GEOJSON = """
+{
+   "type":"MultiPolygon",
+   "crs":{
+      "type":"name",
+      "properties":{
+         "name":"EPSG:3857"
+      }
+   },
+   "coordinates":[
+      [
+         [
+            [1489574.61111389, 6894228.638802719],
+            [89576.815239808, 6894208.60747024],
+            [89576.904295401, 6894207.820852726],
+            [89577.99522641, 6894208.022080451],
+            [89577.961830563, 6894209.229446936],
+            [89589.227363031, 6894210.601454523],
+            [89594.615226386, 6894161.849595264],
+            [89600.314784314, 6894111.37846976],
+            [89651.187791607, 6894116.774968589],
+            [89648.49385993, 6894140.226914071],
+            [89642.92788539, 6894193.423936413],
+            [89639.721884055, 6894224.08372821],
+            [89589.283022777, 6894218.431048969],
+            [89588.192091767, 6894230.248628867],
+            [89574.61111389, 6894228.638802719]
+         ],
+         [
+            [1489610.344670435, 6894182.466199101],
+            [89625.985058891, 6894184.258949757],
+            [89629.547282597, 6894153.270030369],
+            [89613.918026089, 6894151.458993318],
+            [89610.344670435, 6894182.466199101]
+         ]
+      ]
+   ]
+}"""
+
+SAMPLE_POINT_GEOJSON = '{"type":"Point","coordinates":[1.2, 3.4]}'
+
+
+@pytest.fixture
+def shapely_conn(conn):
+    from psycopg.types.shapely import register_shapely
+
+    info = TypeInfo.fetch(conn, "geometry")
+    register_shapely(info, conn)
+    return conn
+
+
+def test_no_adapter(conn):
+    from shapely.geometry import Point
+
+    point = Point(1.2, 3.4)
+    with pytest.raises(ProgrammingError, match="cannot adapt type Point"):
+        conn.execute("SELECT pg_typeof(%s)", [point]).fetchone()[0]
+
+
+def test_with_adapter(shapely_conn):
+    from shapely.geometry import Point, Polygon
+
+    SAMPLE_POINT = Point(1.2, 3.4)
+    SAMPLE_POLYGON = Polygon([(0, 0), (1, 1), (1, 0)])
+
+    assert (
+        shapely_conn.execute(
+            "SELECT pg_typeof(%s)",
+            [SAMPLE_POINT],
+        ).fetchone()[0]
+        == "geometry"
+    )
+
+    assert (
+        shapely_conn.execute(
+            "SELECT pg_typeof(%s)",
+            [SAMPLE_POLYGON],
+        ).fetchone()[0]
+        == "geometry"
+    )
+
+
+@pytest.mark.parametrize("fmt_out", [Format.TEXT, Format.BINARY])
+def test_write_read_shape(shapely_conn, fmt_out):
+    from shapely.geometry import Point, Polygon
+
+    SAMPLE_POINT = Point(1.2, 3.4)
+    SAMPLE_POLYGON = Polygon([(0, 0), (1, 1), (1, 0)])
+
+    fmt_placeholder = "%b" if fmt_out == Format.BINARY else "%t"
```

Comment:
For better coverage of the param format combination you can use a `fmt_in` passing `adapt.Format` as parameter. See for instance:

https://github.com/psycopg/psycopg/blob/7fe7d90c4c35054636cf7b0be18ea410037e5ad5/tests/types/test_uuid.py#L12-L17

Explanation: [`PyFormat`](https://www.psycopg.org/psycopg3/docs/api/adapt.html#psycopg.adapt.PyFormat) is an enum whose values (`s`, `b`, `t`) are the placeholders letters. This is unlike [`pq.Format`](https://www.psycopg.org/psycopg3/docs/api/pq.html#psycopg.pq.Format) which is only "text" and "binary".

Using both `fmt_in` and `fmt_out` you test all the possible 6 combinations, although maybe it's more appropriate to test the dumpers with the `fmt_in` range and the loaders with the `fmt_out` one. The [uuid test](https://github.com/psycopg/psycopg/blob/master/tests/types/test_uuid.py) is a good sample.

Note that I've just simplified the formats iteration in many tests: take a look at the current state of master now.

Reply:
Nice, done!